File: java/src/main/java/io/documentdb/langchain4j/DocumentDBCommandBuilder.java
package io.documentdb.langchain4j;

import dev.langchain4j.store.embedding.EmbeddingSearchRequest;
import java.util.ArrayList;
import java.util.List;
import org.bson.Document;

final class DocumentDBCommandBuilder {
  private DocumentDBCommandBuilder() {}

  static Document createVectorIndex(DocumentDBEmbeddingStoreConfig config) {
    Document options =
        new Document("kind", config.indexKind.wireName())
            .append("dimensions", config.dimensions)
            .append("similarity", config.similarity.name());
    switch (config.indexKind) {
      case IVF -> options.append("numLists", config.ivfNumLists);
      case HNSW ->
          options.append("m", config.hnswM).append("efConstruction", config.hnswEfConstruction);
      case DISKANN ->
          options
              .append("maxDegree", config.diskAnnMaxDegree)
              .append("lBuild", config.diskAnnBuildCandidates);
    }

    Document index =
        new Document("name", config.indexName)
            .append("key", new Document(config.embeddingKey, "cosmosSearch"))
            .append("cosmosSearchOptions", options);
    return new Document("createIndexes", config.collectionName).append("indexes", List.of(index));
  }

  static List<Document> createSearchPipeline(
      DocumentDBEmbeddingStoreConfig config, EmbeddingSearchRequest request) {
    if (request.queryEmbedding() == null) {
      throw new IllegalArgumentException("queryEmbedding is required");
    }
    validateVector(request.queryEmbedding().vector(), config.dimensions, "queryEmbedding");

    Document search =
        new Document("vector", request.queryEmbedding().vectorAsList())
            .append("path", config.embeddingKey)
            .append("k", request.maxResults());
    if (request.filter() != null) {
      search.append("filter", DocumentDBFilterMapper.toBson(request.filter(), config.metadataKey));
    }
    switch (config.indexKind) {
      case IVF -> search.append("nProbes", config.ivfNumProbes);
      case HNSW -> search.append("efSearch", config.hnswEfSearch);
      case DISKANN -> search.append("lSearch", config.diskAnnSearchCandidates);
    }

    List<Document> pipeline = new ArrayList<>();
    pipeline.add(
        new Document(
            "$search", new Document("cosmosSearch", search).append("returnStoredSource", true)));
    pipeline.add(
        new Document(
            "$project",
            new Document("similarityScore", new Document("$meta", "searchScore"))
                .append("document", "$$ROOT")));
    if (request.minScore() > 0) {
      pipeline.add(
          new Document(
              "$match", new Document("similarityScore", new Document("$gte", request.minScore()))));
    }
    pipeline.add(new Document("$limit", request.maxResults()));
    return pipeline;
  }

  static void validateVector(float[] vector, int dimensions, String name) {
    if (vector.length != dimensions) {
      throw new IllegalArgumentException(name + " must contain exactly " + dimensions + " values");
    }
    for (float value : vector) {
      if (!Float.isFinite(value)) {
        throw new IllegalArgumentException(name + " must contain only finite values");
      }
    }
  }
}

File: csharp/src/DocumentDB.LangChain/DocumentDBCommandBuilder.cs
using Microsoft.Extensions.VectorData;
using MongoDB.Bson;

namespace DocumentDB.LangChain;

internal static class DocumentDBCommandBuilder
{
    public static BsonDocument CreateVectorIndex(
        string collectionName,
        VectorMapping vector,
        DocumentDBVectorStoreOptions options)
    {
        var indexKind = vector.Definition.IndexKind ?? IndexKind.DiskAnn;
        var indexOptions = new BsonDocument
        {
            { "kind", ToDocumentDBIndexKind(indexKind) },
            { "dimensions", vector.Definition.Dimensions },
            { "similarity", ToDocumentDBSimilarity(vector.Definition.DistanceFunction) },
        };

        switch (indexKind)
        {
            case IndexKind.IvfFlat:
                indexOptions["numLists"] = options.IvfNumLists;
                break;
            case IndexKind.Hnsw:
                indexOptions["m"] = options.HnswM;
                indexOptions["efConstruction"] = options.HnswEfConstruction;
                break;
            case IndexKind.DiskAnn:
                indexOptions["maxDegree"] = options.DiskAnnMaxDegree;
                indexOptions["lBuild"] = options.DiskAnnBuildCandidates;
                break;
            default:
                throw new NotSupportedException($"DocumentDB does not support index kind '{indexKind}'.");
        }

        return new BsonDocument
        {
            { "createIndexes", collectionName },
            {
                "indexes",
                new BsonArray
                {
                    new BsonDocument
                    {
                        { "name", GetIndexName(vector.StorageName) },
                        { "key", new BsonDocument(vector.StorageName, "cosmosSearch") },
                        { "cosmosSearchOptions", indexOptions },
                    },
                }
            },
        };
    }

    public static BsonDocument[] CreateSearchPipeline(
        VectorMapping vector,
        ReadOnlyMemory<float> query,
        int top,
        int skip,
        double? scoreThreshold,
        BsonDocument? filter,
        DocumentDBVectorStoreOptions options)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(top, 1);
        ArgumentOutOfRangeException.ThrowIfNegative(skip);
        if (query.IsEmpty)
        {
            throw new ArgumentException("The query vector must not be empty.", nameof(query));
        }

        var search = new BsonDocument
        {
            { "vector", new BsonArray(query.Span.ToArray()) },
            { "path", vector.StorageName },
            { "k", checked(top + skip) },
        };
        if (filter is not null)
        {
            search["filter"] = filter;
        }

        switch (vector.Definition.IndexKind ?? IndexKind.DiskAnn)
        {
            case IndexKind.IvfFlat:
                search["nProbes"] = options.IvfNumProbes;
                break;
            case IndexKind.Hnsw:
                search["efSearch"] = options.HnswEfSearch;
                break;
            case IndexKind.DiskAnn:
                search["lSearch"] = options.DiskAnnSearchCandidates;
                break;
        }

        var pipeline = new List<BsonDocument>
        {
            new("$search", new BsonDocument
            {
                { "cosmosSearch", search },
                { "returnStoredSource", true },
            }),
            new("$project", new BsonDocument
            {
                { "similarityScore", new BsonDocument("$meta", "searchScore") },
                { "document", "$$ROOT" },
            }),
        };

        if (scoreThreshold is { } threshold)
        {
            pipeline.Add(new BsonDocument("$match", new BsonDocument(
                "similarityScore",
                new BsonDocument("$gte", threshold))));
        }

        if (skip > 0)
        {
            pipeline.Add(new BsonDocument("$skip", skip));
        }

        pipeline.Add(new BsonDocument("$limit", top));
        return pipeline.ToArray();
    }

    public static string GetIndexName(string storageName) =>
        $"{storageName.Replace(".", "_", StringComparison.Ordinal)}_vector";

    private static string ToDocumentDBIndexKind(string indexKind) => indexKind switch
    {
        IndexKind.IvfFlat => "vector-ivf",
        IndexKind.Hnsw => "vector-hnsw",
        IndexKind.DiskAnn => "vector-diskann",
        _ => throw new NotSupportedException($"DocumentDB does not support index kind '{indexKind}'."),
    };

    private static string ToDocumentDBSimilarity(string? distanceFunction) => distanceFunction switch
    {
        null or DistanceFunction.CosineSimilarity or DistanceFunction.CosineDistance => "COS",
        DistanceFunction.EuclideanDistance or DistanceFunction.EuclideanSquaredDistance => "L2",
        DistanceFunction.DotProductSimilarity or DistanceFunction.NegativeDotProductSimilarity => "IP",
        _ => throw new NotSupportedException($"DocumentDB does not support distance function '{distanceFunction}'."),
    };
}